test/WasabiBNPL.test.ts

```diff
@@ -111,13 +121,112 @@ contract("WasabiBNPL", (accounts) => {
       [testNft.address, tokenToBuy.toString(), loanAmount, repayment]
     );
 
-    const res = await bnpl.bnpl(
+    // List of marketplace calldata and Signature's length is not same, should revert.
+    await truffleAssert.reverts(
+      bnpl.bnpl(
+        nftLending.address,
+        borrowData,
+        toEth(13),
+        [buyCall],
+        [],
+        metadata(buyer, 3.5)
+      ),
+      "Length is invalid"
+    );
+
+    // List of marketplace calldata's length can't be zero, should revert.
+    await truffleAssert.reverts(
+      bnpl.bnpl(
+        nftLending.address,
+        borrowData,
+        toEth(13),
+        [],
+        signatures,
+        metadata(buyer, 3.5)
+      ),
+      "Need marketplace calls"
+    );
+
+    // When trying to sign with buyer address, should revert.
+    const wrongSignature = await signFunctionCallData(buyCall, buyer);
+
+    await truffleAssert.reverts(
+      bnpl.bnpl(
+        nftLending.address,
+        borrowData,
+        toEth(13),
+        [buyCall],
+        [wrongSignature],
+        metadata(buyer, 3.5)
+      ),
+      "Owner is not signer"
+    );
+
+    const invalidLendingAddress = "0xb0d1140a09f669935b4848f6826fd16ff19787b9";
+
+    // Nft Lending address should be added in AddressProvider, should revert.
+    await expectRevertCustomError(
+      bnpl.bnpl(
+        invalidLendingAddress,
+        borrowData,
+        toEth(13),
+        [buyCall],
+        signatures,
+        metadata(buyer, 3.5)
+      ),
+      "InvalidParam",
+      "Nft Lending Address is invalid."
+    );
+
+    optionId = await bnpl.bnpl.call(
+      nftLending.address,
+      borrowData,
+      toEth(13),
+      [buyCall],
+      signatures,
+      metadata(buyer, 3.5)
+    );
+
+    await bnpl.bnpl(
       nftLending.address,
       borrowData,
       toEth(13),
       [buyCall],
       signatures,
       metadata(buyer, 3.5)
     );
+
+    assert.equal(await option.ownerOf(optionId), buyer);
+  });
+
+  it("should get option data", async () => {
+    let optionData = await bnpl.getOptionData(optionId);
```

Comment:
nice, looking much better!

Can we also add assertions before and after to check that the flash loan contract has won some money